### PR TITLE
feat!: coverage-c8 to use V8 profiler directly instead of `NODE_V8_COVERAGE`

### PR DESCRIPTION
**Pull request by AriPerkkio**
Breaking changes:
- `vitest` and `@vitest/coverage-c8` versions must match as the `CoverageProviderModule` introduces new APIs and c8-provider starts to use these. 
- Removes `CoverageProvider.onBeforeFilesRun`. Custom coverage providers should use  `CoverageProviderModule.startCoverage` instead.

Currently Vitest is using [`NODE_V8_COVERAGE`](https://github.com/nodejs/node/blob/main/doc/api/cli.md#node_v8_coveragedir) environment variable to generate V8 coverage reports for `@vitest/coverage-c8`. We've used this environment variable to trigger Node to start writing JSON reports to `<reports-dir>/tmp`. 

These reports contain coverage results of all run code, including Node's internal `node:<module>` and unnecessary third party `**/node_modules/**` packages. In addition the results contain [`source-map-cache` information](https://github.com/nodejs/node/blob/main/doc/api/cli.md#source-map-cache) which is completely irrelevant for Vitest. There is so much data on these JSON reports. In #2108 user reported **435 MB**  `coverage/tmp` directory on CI. 

Also using `NODE_V8_COVERAGE` with `worker_threads` has some weird features: https://github.com/nodejs/node/issues/46378. I think it's meant more for running a single script instead of to be utilized by test runners. 

This PR replaces `process.env.NODE_V8_COVERAGE` usage by controlling V8 Profiler directly using `node:inspect`. We don't enable unncessary `source-map-cache` and filter the ScriptCoverage results early so the amount of stored data is reduced a lot. Results are stored in memory instead of file system - exactly as `@vitest/coverage-istanbul` does. This reduces amount of file system handles. Since we now know what files are in the report, we can reduce the amount of Vite source map handling in [`provider.ts`](https://github.com/vitest-dev/vitest/blob/4fc492c61d6e01bc34bc807a6b57197a1914f823/packages/coverage-c8/src/provider.ts#L74-L76). 

This should help with flaky coverage test on CI since workers no more write to `<reports-dir>/tmp` when they exit.

### CODE REVIEW

**Comment by AriPerkkio** [comment]
Minimal example for testing Profiler's coverage commands: https://gist.github.com/AriPerkkio/c2df034cd71b0dfcc934c477758c20ee

**Review by sheremet-va — APPROVED** [review]

**Comment by AriPerkkio** [comment]
I did some manual regression testing with `vuejs/core` repository as these changes are quite critical. This seems to make report generation a lot faster now that amount of data is reduced, 3600ms -> 400ms:

```
Main branch:
provider.ts :: reportCoverage: 3.667s
provider.ts :: reportCoverage: 3.526s
provider.ts :: reportCoverage: 3.699s

PR branch:
provider.ts :: reportCoverage: 405.988ms
provider.ts :: reportCoverage: 399.793ms
provider.ts :: reportCoverage: 449.856ms
```


Coverage reports are almost identical. Branch count is off by one but after checking that one manually it doesn't make sense in either case.

```
Main:
=============================== Coverage summary ===============================
Statements   : 94.27% ( 39675/42086 )
Branches     : 92.07% ( 8987/9760 )
Functions    : 91.73% ( 2053/2238 )
Lines        : 94.27% ( 39675/42086 )
================================================================================

PR branch:
=============================== Coverage summary ===============================
Statements   : 94.27% ( 39675/42086 )
Branches     : 92.07% ( 8986/9759 )
Functions    : 91.73% ( 2053/2238 )
Lines        : 94.27% ( 39675/42086 )
================================================================================
```

**Comment by sheremet-va** [comment]
> This seems to make report generation a lot faster now that amount of data is reduced, 3600ms -> 400ms:

These are great improvements 😱 👍🏻

**Comment by AriPerkkio** [comment]
❗ Some manual resolving will be required if #2817 is merged first. Custom provider test should be updated to use the new `CoverageProviderModule`'s APIs as well.

**Comment by sheremet-va** [comment]
This PR has a conflict now since #2817 was fixed first, as you said. 😄

**Comment by AriPerkkio** [comment]
I'll start to work on rebasing `main` here now.

File: test/coverage-test/custom-provider.ts [hunk 1]
```diff
@@ -11,6 +11,10 @@ const CustomCoverageProviderModule: CoverageProviderModule = {
   takeCoverage() {
     return { customCoverage: 'Coverage report passed from workers to main thread' }
   },
+
+  // TODO: Figure out how to verify these were called in tests. Maybe write file to "./coverage" ?
+  startCoverage() {},
+  stopCoverage() {},
```
Comment (AriPerkkio) [inline test/coverage-test/custom-provider.ts:17]: I verified manually that these are indeed called during tests. I'll need to think proper way to include these in tests.

**Comment by AriPerkkio** [comment]
Rebase done, most conflicts were in `integrations/coverage.ts`. Ubuntu+Node18 CI is failing on unrelated test. 
The `test/coverage-test/` tests pass and manual regression with `vuejs/core` gives good results:

<details>

```
# PR branch

 % Coverage report from c8

=============================== Coverage summary ===============================
Statements   : 94.27% ( 39675/42086 )
Branches     : 92.07% ( 8986/9759 )
Functions    : 91.73% ( 2053/2238 )
Lines        : 94.27% ( 39675/42086 )
================================================================================
reportCoverage: 418.102ms


# Main branch
 % Coverage report from c8

=============================== Coverage summary ===============================
Statements   : 94.27% ( 39675/42086 )
Branches     : 92.07% ( 8987/9760 )
Functions    : 91.73% ( 2053/2238 )
Lines        : 94.27% ( 39675/42086 )
================================================================================
reportCoverage: 3.828s
```

</details>

**Pull request merged**